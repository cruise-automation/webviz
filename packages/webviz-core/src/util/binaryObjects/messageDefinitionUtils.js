// @flow
//
//  Copyright (c) 2020-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { sum } from "lodash";
import memoize from "memoize-weak";

import type { RosMsgField, RosDatatypes } from "webviz-core/src/types/RosDatatypes";

const primitiveSizes = {
  string: 8,
  bool: 1,
  int8: 1,
  uint8: 1,
  int16: 2,
  uint16: 2,
  int32: 4,
  uint32: 4,
  float32: 4,
  float64: 8,
  int64: 8,
  uint64: 8,
  time: 8,
  duration: 8,
};

export const typeSize = memoize(
  (typesByName: RosDatatypes, typeName: string): number => {
    if (primitiveSizes[typeName] != null) {
      return primitiveSizes[typeName];
    }
    const messageType = typesByName[typeName];
    if (messageType == null) {
      throw new Error(`Unknown type: ${typeName}`);
    }
    // We add field sizes here, not type sizes. If the type has a field that's an array of strings,
    // it only adds eight bytes to the object's inline size.
    return sum(typesByName[typeName].fields.map((field) => fieldSize(typesByName, field)));
  }
);

// No point in memoizing this -- it can only do O(1) work before hitting `typeSize`, which is
// memoized. Memoizing on the string type name will work better than memoizing on field object
// identity, too.
export function fieldSize(typesByName: RosDatatypes, field: RosMsgField): number {
  if (field.isConstant) {
    return 0;
  }
  if (field.isArray) {
    // NOTE: Fixed-size arrays are not inlined.
    return 2 * primitiveSizes.int32;
  }
  return typeSize(typesByName, field.type);
}

// It's useful for field-accessor code for times and durations to share codepaths with field
// accessor code for complex message types. To do that, we add them to the set of messages.
// (Note, this might change depending on the code written next.)
export const addTimeTypes = (typesByName: RosDatatypes): RosDatatypes => ({
  ...typesByName,
  time: { fields: [{ name: "sec", type: "int32" }, { name: "nsec", type: "int32" }] },
  duration: { fields: [{ name: "sec", type: "int32" }, { name: "nsec", type: "int32" }] },
});

export const friendlyTypeName = (name: string): string => name.replace("/", "_");
