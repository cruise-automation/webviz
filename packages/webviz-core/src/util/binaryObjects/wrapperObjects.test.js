// @flow
//
//  Copyright (c) 2020-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { definitions } from "./messageDefinitionUtils.test";
import {
  getGetClassesForView,
  printFieldDefinition,
  printGetClassesForView,
  printSingularExpression,
  PointerExpression,
} from "./wrapperObjects";

describe("PointerExpression", () => {
  it("prints a constructed expression nicely", () => {
    expect(new PointerExpression("this.offset").toString()).toBe("this.offset");
  });

  it("adds correctly", () => {
    expect(new PointerExpression("this.offset").add(10).toString()).toBe("(this.offset + 10)");
    expect(
      new PointerExpression("this.offset")
        .add(10)
        .add(-10)
        .toString()
    ).toBe("this.offset");
  });
});

describe("printSingularExpression", () => {
  it("handles strings", () => {
    const pointer = new PointerExpression("this.offset").add(1);
    expect(printSingularExpression({}, "string", pointer)).toBe(
      "$bigString.slice($view.getInt32((this.offset + 1), true), $view.getInt32((this.offset + 5), true))"
    );
  });

  it("handles times", () => {
    const pointer = new PointerExpression("this.offset");
    expect(printSingularExpression(definitions, "time", pointer)).toBe("new time(this.offset)");
  });

  it("handles primitives", () => {
    const pointer = new PointerExpression("this.offset");
    expect(printSingularExpression(definitions, "int32", pointer)).toBe("$view.getInt32(this.offset, true)");
    expect(printSingularExpression(definitions, "uint64", pointer)).toBe("$view.getFloat64(this.offset, true)");
    expect(printSingularExpression(definitions, "bool", pointer)).toBe("($view.getUint8(this.offset) !== 0)");
  });

  it("handles complex types", () => {
    const pointer = new PointerExpression("this.offset");
    expect(printSingularExpression(definitions, "std_msgs/Header", pointer)).toBe("new std_msgs_Header(this.offset)");
  });

  it("throws on unknown types", () => {
    const pointer = new PointerExpression("this.offset");
    expect(() => printSingularExpression(definitions, "not_a/Type", pointer)).toThrow("unknown type");
  });
});

describe("printFieldDefinition", () => {
  it("handles complex arrays", () => {
    const pointer = new PointerExpression("this.offset");
    const field = definitions["fake_msgs/HasComplexArray"].fields[0];
    expect(printFieldDefinition(definitions, field, pointer)).toBe(
      `
complexArray() {
  const from = $view.getInt32(this.offset, true);
  const to = $view.getInt32((this.offset + 4), true);
  return new fake_msgs_HasComplexAndArray$Array(from, to);
}`.trim()
    );
  });

  it("special-cases byte arrays", () => {
    const pointer = new PointerExpression("this.offset");
    const field = definitions["fake_msgs/HasByteArray"].fields[0];
    expect(printFieldDefinition(definitions, field, pointer)).toBe(
      `
byte_array() {
  const from = $view.getInt32(this.offset, true);
  const to = $view.getInt32((this.offset + 4), true);
  return new Uint8Array($view.buffer, from, to - from);
}`.trim()
    );
  });

  it("handles singular fields", () => {
    const pointer = new PointerExpression("this.offset");
    const field = definitions["std_msgs/Header"].fields[0];
    expect(printFieldDefinition(definitions, field, pointer)).toBe(
      `
seq() {
  return $view.getUint32(this.offset, true);
}`.trim()
    );
  });

  it("makes a static class function for constants", () => {
    const pointer = new PointerExpression("this.offset");
    const field = definitions["fake_msgs/HasConstant"].fields[0];
    expect(printFieldDefinition(definitions, field, pointer)).toBe(
      `
static const() {
  return 1;
}`.trim()
    );
  });
});

describe("printGetClassesForView", () => {
  it("returns the expected code", () => {
    expect(printGetClassesForView(definitions)).toMatchSnapshot();
  });
});

describe("getGetClassesForView", () => {
  it("returns an object", () => {
    const intArray = new Int32Array([12, 34]);
    const view = new DataView(intArray.buffer);
    const bigString = "";
    // Mostly just check that the code compiles for now.
    const classes = getGetClassesForView(definitions)(view, bigString);
    expect(classes).toEqual({
      "std_msgs/Header": expect.any(Function),
      "fake_msgs/HasComplexAndArray": expect.any(Function),
      "fake_msgs/HasComplexArray": expect.any(Function),
      "fake_msgs/HasConstant": expect.any(Function),
      "fake_msgs/HasByteArray": expect.any(Function),
      duration: expect.any(Function),
      time: expect.any(Function),
    });
    const t = new classes.time(0);
    expect(t.sec()).toBe(12);
    expect(t.nsec()).toBe(34);

    expect(classes["fake_msgs/HasConstant"].const()).toBe(1);
  });
  // TODO: Test arrays, strings etc.
});
