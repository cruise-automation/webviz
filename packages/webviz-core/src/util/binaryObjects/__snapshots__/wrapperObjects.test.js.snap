// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printGetClassesForView returns the expected code 1`] = `
"const $offset = Symbol();
class std_msgs_Header {
  constructor(offset) {
    this[$offset] = offset;
  }
  seq() {
    return $view.getUint32(this[$offset], true);
  }
  stamp() {
    return new time((this[$offset] + 4));
  }
  frame_id() {
    return $bigString.slice($view.getInt32((this[$offset] + 12), true), $view.getInt32((this[$offset] + 16), true));
  }
}
class fake_msgs_HasComplexAndArray {
  constructor(offset) {
    this[$offset] = offset;
  }
  header() {
    return new std_msgs_Header(this[$offset]);
  }
  stringArray() {
    const from = $view.getInt32((this[$offset] + 20), true);
    const to = $view.getInt32((this[$offset] + 24), true);
    return new string$Array(from, to);
  }
}
class fake_msgs_HasComplexArray {
  constructor(offset) {
    this[$offset] = offset;
  }
  complexArray() {
    const from = $view.getInt32(this[$offset], true);
    const to = $view.getInt32((this[$offset] + 4), true);
    return new fake_msgs_HasComplexAndArray$Array(from, to);
  }
}
class fake_msgs_HasConstant {
  constructor(offset) {
    this[$offset] = offset;
  }
  static const() {
    return 1;
  }
}
class fake_msgs_HasByteArray {
  constructor(offset) {
    this[$offset] = offset;
  }
  byte_array() {
    const from = $view.getInt32(this[$offset], true);
    const to = $view.getInt32((this[$offset] + 4), true);
    return new Uint8Array($view.buffer, from, to - from);
  }
}
class time {
  constructor(offset) {
    this[$offset] = offset;
  }
  sec() {
    return $view.getInt32(this[$offset], true);
  }
  nsec() {
    return $view.getInt32((this[$offset] + 4), true);
  }
}
class duration {
  constructor(offset) {
    this[$offset] = offset;
  }
  sec() {
    return $view.getInt32(this[$offset], true);
  }
  nsec() {
    return $view.getInt32((this[$offset] + 4), true);
  }
}
return {
  \\"std_msgs/Header\\": std_msgs_Header,
  \\"fake_msgs/HasComplexAndArray\\": fake_msgs_HasComplexAndArray,
  \\"fake_msgs/HasComplexArray\\": fake_msgs_HasComplexArray,
  \\"fake_msgs/HasConstant\\": fake_msgs_HasConstant,
  \\"fake_msgs/HasByteArray\\": fake_msgs_HasByteArray,
  \\"time\\": time,
  \\"duration\\": duration,
};"
`;
